### DISCUSSION TITLE
Understanding the behavior of `generateStaticParams()` when it returns an empty array

### DISCUSSION THREAD

**Discussion by ntopia** [vercel/next.js#60621] [Help]
### Summary

I've encountered an unusual behavior with the `generateStaticParams` function.

Specifically, the function seems to cache its responses of dynamic routes when `generateStaticParams` is utilized without returning any valid values. This observation led me to a question, as the [documentation](https://nextjs.org/docs/app/api-reference/functions/generate-static-params) doesn't provide details about this specific scenario: What happens when `generateStaticParams` returns an empty array?

To illustrate this, I've prepared two reproducible examples:

The first example does not use `generateStaticParams`: [sandbox link](https://codesandbox.io/p/devbox/grwb69w87df5-forked-6x94c3?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrd5mrjb00073b6h7isy0fvc%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrd5mrjb00023b6h8m02g8t7%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrd5mrjb00043b6h3pk6agu3%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrd5mrjb00063b6hl5fgxudu%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60.94606733498616%252C39.05393266501384%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrd5mrjb00023b6h8m02g8t7%2522%253A%257B%2522id%2522%253A%2522clrd5mrjb00023b6h8m02g8t7%2522%252C%2522activeTabId%2522%253A%2522clrd5ziwj027d3b6hfna3mveb%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522clrd5ziwj027d3b6hfna3mveb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522clrd5mrjb00063b6hl5fgxudu%2522%253A%257B%2522id%2522%253A%2522clrd5mrjb00063b6hl5fgxudu%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clrd5yw2v021j3b6h2zs6dadb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252Fshow%252F3%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clrd5yw2v021j3b6h2zs6dadb%2522%257D%252C%2522clrd5mrjb00043b6h3pk6agu3%2522%253A%257B%2522id%2522%253A%2522clrd5mrjb00043b6h3pk6agu3%2522%252C%2522activeTabId%2522%253A%2522clrd64hvj005d3b6he32qr8vq%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522id%2522%253A%2522clrd64hvj005d3b6he32qr8vq%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

The second example uses `generateStaticParams`, but it returns an empty array: [sandbox link](https://codesandbox.io/p/devbox/grwb69w87df5-pyldnl?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrd5mrjb00073b6h7isy0fvc%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrd5mrjb00023b6h8m02g8t7%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrd5mrjb00043b6h3pk6agu3%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrd5mrjb00063b6hl5fgxudu%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B61.957888685279%252C38.042111314721%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrd5mrjb00023b6h8m02g8t7%2522%253A%257B%2522id%2522%253A%2522clrd5mrjb00023b6h8m02g8t7%2522%252C%2522activeTabId%2522%253A%2522clrd5ziwj027d3b6hfna3mveb%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftasks.json%2522%252C%2522id%2522%253A%2522clrd5ziwj027d3b6hfna3mveb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522clrd5mrjb00063b6hl5fgxudu%2522%253A%257B%2522id%2522%253A%2522clrd5mrjb00063b6hl5fgxudu%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clrd5yw2v021j3b6h2zs6dadb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252Fshow%252F3%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clrd5yw2v021j3b6h2zs6dadb%2522%257D%252C%2522clrd5mrjb00043b6h3pk6agu3%2522%253A%257B%2522id%2522%253A%2522clrd5mrjb00043b6h3pk6agu3%2522%252C%2522activeTabId%2522%253A%2522clrd62p7i003x3b6ho233qzd9%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522id%2522%253A%2522clrd62p7i003x3b6ho233qzd9%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Both examples can be observed by accessing the URL `/show/3`. In the first case, each page refresh logs a "hit" message. However, in the second case, the "hit" message appears only twice, and subsequent refreshes produce no log, even though the page is refreshed multiple times.

The notable difference in these examples lies in the `app/show/[id]/page.tsx` file. In one instance, `generateStaticParams` is absent, while in the other, it returns an empty array.

Could you help me understand the rationale behind this behavior?

Other information:
- Next.js 14.0.4
- Command: `pnpm build && pnpm start`  (`pnpm dev` doesn't show this behavior)

**Comment by icyJoseph** [comment]
I think, without generateStaticParams, is more akin to GetServerSideProps, that is, each page is generated on demand.

When you add generateStaticParams, is more akin to ISR (GetStaticProps + GetStaticPaths), where you could pre-render a bunch of pages, or none, and then let the ISR generate them during the runtime, or also generate a few more. You could've also let it open so that it pre-rendered a bunch of pages during build time, and none during runtime. 

One example was, a shop with thousands of products, you could pre-render the top 10% by some metric, say page views, so that they have near zero serve time, and the other 90% that's not visited as much, could be generated when a `next/link` points to them, or when users require them.

Notably, these results are cached, as pre-rendered pages. They can be revalidated as well.

GetServerSideProps in the other hand, does a render pass every time the page is requested.

I remember doing some tests on App Router around November, and there was still not 100% match between Pages Router ISR and App Router though. I think the fallback variable was not really working correctly.
